tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,231 @@
+    POIntegrationPage,
+    POOverridenPage,
+    Product,
+    ProductFromInjectable,
+    ProductFromInjectablePage,
+    ProductPage,
+    ReplacedProduct,
+    ReplacedProductPage,
+    StandaloneProduct,
+    StandaloneProductPage,
+    SubclassProductPage,
+    SubclassReplacedProduct,
+    SubclassReplacedProductPage,
+)
+from tests.test_middleware import ProductHtml, spider_for
+from tests.utils import crawl_single_item, create_scrapy_settings
+
+# The rules are defined in `tests/po_lib/__init__.py`.
+RULES = default_registry.get_rules()
+
+
+def rules_settings() -> dict:
+    settings = create_scrapy_settings(None)
+    # Converting it to a set removes potential duplicate ApplyRules
+    settings["SCRAPY_POET_OVERRIDES"] = set(RULES)
+    return settings
+
+
+@inlineCallbacks
+def crawl_item(PageObject) -> Any:
+    """Helper function to easily return the item from a simulated Scrapy callback
+    which asks for either of these dependencies:
+        - Page Object
+        - Item Type
+    """
+    item, _, _ = yield crawl_single_item(
+        spider_for(PageObject), ProductHtml, rules_settings(), port=PORT
+    )
+    return item
+
+
+@inlineCallbacks
+def test_overrides() -> None:
+    """This tests the rules where a PageObject is used in lieu of another."""
+    item = yield crawl_item(POOverridenPage)
+    assert item == {"msg": "PO replacement"}
+
+
+@inlineCallbacks
+def test_item_return() -> None:
+    """This tests the rules where an Item Type is requested.
+
+    A corresponding ``ApplyRule`` should be available to denote that a Page
+    Object is equipped to produce it for the given URL pattern and Item Type.
+
+    For example, the code below should return a ``Product`` item which was taken
+    using the route ``Returns[ItemType]``.
+
+    ..code-block::
+
+        @handle_urls(URL)
+        class ProductPage(ItemPage[Product]):
+            ...
+    """
+    item = yield crawl_item(Product)
+    assert item == Product(name="product's name")
+
+
+@inlineCallbacks
+def test_item_return_subclass() -> None:
+    """A Page Object should properly inherit the ``Return[ItemType]`` from its
+    parent.
+
+    For the example below, asking for ParentProduct would return ``ParentProduct``
+    from the ``ParentProductPage`` since the rules have equal priority but the
+    ``ParentProductPage`` has been decleared first.
+
+    ..code-block::
+
+        @handle_urls(URL)
+        class ParentProductPage(ItemPage[ParentProduct]):
+            ...
+
+        @handle_urls(URL)
+        class SubclassProductPage(ParentProductPage):
+            ...
+
+    If users want to use the ``ParentProduct`` coming from ``SubclassProductPage``,
+    then they'd need to increase the priority:
+
+    ..code-block::
+
+        @handle_urls(URL, priority=600)
+        class SubclassProductPage(ParentProductPage):
+            ...
+
+    """
+    item = yield crawl_item(ParentProduct)
+    assert item == ParentProduct(name="subclass product's name")
+
+
```

Comment:
I wanted to discuss the situation with the three (3) test cases below which are currently failing, as they try to abide the idea from https://github.com/scrapinghub/web-poet/pull/84#discussion_r985853698:
> The idea is that `to_return` passed explicitly to the decorator should take priority over the item class defined by the page object.

We could handle this in **scrapy-poet** by replacing the original item returned by the Page Object with the item class specified in `to_return`. Some examples include:

a. `dict` to item class

```python
# using the rule should return StandaloneProduct
@handle_urls(URL, to_return=StandaloneProduct)
class StandaloneProductPage(ItemPage):
    ... # to_item() returns a dict
```

b. an item class into another item class (typically because of subclasses)

```python
@handle_urls(URL)
class ParentReplacedProductPage(ItemPage[ParentReplacedProduct]):
    ...

# using the rule should return StandaloneProdSubclassReplacedProductuct
@handle_urls(URL, to_return=SubclassReplacedProduct) 
class SubclassReplacedProductPage(ParentReplacedProductPage):
    ... # to_item() still returns ParentReplacedProduct
```

---

However, this comes with some changes to **web-poet**:

### 1. `@handle_urls` may need to be updated to also support `skip_nonitem_fields`

This ensures that the current API in **web-poet** (https://web-poet.readthedocs.io/en/stable/advanced/fields.html#changing-item-type) below:

```python
@attrs.define
class CustomItem:
    name: str

class CustomFooPage(FooPage, Returns[CustomItem], skip_nonitem_fields=True):
    pass
```

is compatible with *changing the item type* with **web-poet**'s `@handle_urls` as well:

```python
@attrs.define
class CustomItem:
    name: str

@handle_urls(URL, returns=CustomItem, skip_nonitem_fields=True)
class CustomFooPage(FooPage):
    pass
```

They do different things:

- changing the return item type directly from the Page Object.
- changing the return item type when using the rules for a given Page Object.

-----

There's also some slight implementation challenges as well:

### 2. Converting objects into another item class

The object to be converted could be instances of `dict`, `dataclasses`, `attrs`, etc. depending on how the developer creates them.

Converting `dict` is straighforward as `item_cls(**item_dict)`.

Converting `dataclasses` and `attrs` instances could use `dataclass.asdict` and `attrs.asdict` respectively. For example:

```python
@attrs.define
class TargetItem:
    x: str
    y: int

@attrs.define
class AttrsItem:
    x: str
    y: int
        
    def do_something(self):
        print("action")

attrs_item = AttrsItem(x="data", y=1)

TargetItem(**attrs.asdict(attrs_item))
# TargetItem(x='data', y=1)
```

The challenge here is when developers use their own classes which makes it hard to determine which are the actual attributes containing data vs helper methods. Item classes should make it easy to derive such things for easier item conversion.

```python
class CustomItem:
    def __init__(self, x: int, y: str):
        self._x = x
        self._y = y
        
    # ... some setters and getters here that act similarly with dataclasses/attrs
        
    @property
    def x(self):
        return self._x
    
    @property
    def y(self):
        return self._y
        
    def do_something(self):
        print("action")
    
custom_item = CustomItem(x="data", y=1)

dir(custom_item)
# [..., 'do_something', 'x', 'y']
```

Some approaches we could do:
- enforce a requirement for items return by Page Objects to be either instances of `dict`, `dataclass`, or `attrs`.
- introduce a parameter which accepts a `Callable` that handles the item conversion. The developer could use this to make their custom items work.

Reply:
See https://github.com/scrapinghub/scrapy-poet/pull/88#discussion_r993492293.